### ISSUE TITLE
Expose a method to query for a description of an interface

### ISSUE THREAD

**Issue by MitchRudominer** [domokit/mojo#448]
We now have structures that can describe Mojom types.
https://github.com/domokit/mojo/blob/master/mojo/public/interfaces/bindings/mojom_descriptors.mojom

Next we would like a way to query for a description of an interface and get back one of these structures (or something similar.)

One possible way to implement this is to add a new "command" to the Run() method of interface control messages
https://github.com/domokit/mojo/blob/master/mojo/public/interfaces/bindings/interface_control_messages.mojom

**Comment by MitchRudominer** [comment]
If we end up going with the strategy I describe above of adding a new command to the Run() method of interface control messages, then in order to implement in Go we need to first fix #447.

**Comment by jamesr** [comment]
Control messages are really tricky unless the receiver always knows when to expect them since they break write->wait->read. I would recommend defining a separate interface for queries and say that developers need to expose this interface if they wish a particular interface be dynamically queryable.

**Comment by jamesr** [comment]
Note that the current control messages are sent before any other messages.

**Comment by yzshen** [comment]
Those control messages are sent because the user makes calls on InterfacePtr. So they are not very different from other "normal" messages.

(Not saying that it is better than "defining a separate interface", though.)

**Comment by MitchRudominer** [comment]
One advantage of control messages is that they avoid the problem of needing to identify the interface to be described by name. Suppose two parties share a message pipe associated to an interface but that connection was not initiated via ServiceProvider.ConnectToService(). Suppose party one wants to ask party two to describe the interface. If the mechanism to do so is some method on some other interface then there needs to be a way to say "describe the interface from that other pipe."

Or we can say that we only provide builtin support for describing interfaces obtained from ServiceProvider.ConnectToService().

**Comment by MitchRudominer** [comment]
James will you please flesh out your concern about write->wait->read. Can you please give a concrete example of a problematic interaction. Just trying to understand your concern.

**Comment by jamesr** [comment]
Given a mojom interface like this:

interface Foo {
  Call() => (bool happy);
};

and a MojoHandle |foo| bound to an implementation of the interface |Foo| if my code does this:

MojoWriteMessage(foo, [Call message]);
MojoWait(foo, MOJO_HANDLE_SIGNAL_READABLE | MOJO_HANDLE_SIGNAL_PEER_CLOSED, ...);
MojoReadMessage(foo, ...);

then my code should know the value of |happy| in the response unless the pipe closed.  If the other side of the pipe has the ability to generate incoming messages then the wait could return for reasons other than a response to my outgoing call.  Similarly if my code does not have an outstanding call to |Call| then it should be able to proceed without waiting on |foo| at all.

**Comment by yzshen** [comment]
The existing control message requests are initiated by the user and from the same side of the message pipe as the interface method requests. So as long as there is no pending control message request waiting for response (which the user can decide because he/she initiated the control message requests, if any), the above pattern still works.

**Comment by jamesr** [comment]
Right, but for an interface query this would mean the only party that could ask for the shape of an interface is the side that already knows it which isn't as useful.  There needs to be a way for the initiating side to ask for the other side's description of an interface.

**Comment by yzshen** [comment]
Isn't it the "user" side of the interface who wants to query for interface description?
If yes, then those interface query requests will flow in the same direction as the normal method requests and existing control message requests, right?

(I am probably out of context and have missed something. Just trying to provide info about my previous work. Please ignore if I am totally wrong. :))

**Comment by MitchRudominer** [comment]
The use case I have in mind is as Yuzhu says: The client side of the
interface is the one who wants to query for the interface description and
the implementor side of the interface will provide the description. This is
what I was considering using control messages for.

There is also a different use case in which the client side will provide a
description of the interface but I didn't have in mind to use control
messages for that. That could be done in a couple of different ways such as
when the client first initiates the connection he gives the description of
the interface at the same time.

**Comment by MitchRudominer** [comment]
It seems that there is not much support for using control messages as a means of exposing interface descriptions. For my latest thinking about how to expose interface descriptions see
https://codereview.chromium.org/1381023005/.

**Comment by MitchRudominer** [comment]
This didn't link properly:
https://codereview.chromium.org/1719873003/
